### PR TITLE
Fix remaining casting issue when scaling compressed fits files

### LINKED ISSUE #4600: Exception for FITS extensions where BZERO cannot be implicitly cast to same type as data
Calling `img.scale(bzero=bzero)`, where `img` is an _ImageBaseHDU object fails with a TypeError when `bzero` is non-zero and has a 'higher' type than `img.data`. The [offending line](https://github.com/astropy/astropy/blob/master/astropy/io/fits/hdu/image.py#L442) attempts to add a float to an integer array in-place.  A pull request will be issued and associated with this issue shortly.

The similar code in _ImageBaseHDU._get_scaled_image_data does not suffer from this problem, as the data is cast up to float before applying the scaling.

The FITS standard says that BZERO [should be a float](http://heasarc.gsfc.nasa.gov/docs/fcg/standard_dict.html), so it is not appropriate to change its type to match the data during verification. 

This issue was encountered with in astropy-1.1.1 on Python-2.7.11, when attempting to run `solve_field` from [astrometry.net](https://github.com/dstndstn/astrometry.net)-0.63 by @dstndstn.  The problem is [still present](https://github.com/dstndstn/astrometry.net/blob/master/util/fits2fits.py#L86) in the astrometry.net master branch  It prevents `solve_field` from working on images with integer data and non-zero BZERO.

### PR DESCRIPTION
**Pull request by mwcraig**
This fixes the issue addressed in #4639 that were not fixed in #5250. Essentially a squash with edits as needed to resolve conflicts with #5250 and remove the parts of #4639 that had been fixed by #5250. 

Closing #4639 and #4602 (an alternative to #4639).

Fixes #4600.

### CODE REVIEW

File: astropy/io/fits/tests/test_image.py [hunk 1]
```diff
@@ -1550,14 +1590,42 @@ def test_duplicate_compression_header_keywords(self):
...
+
+        hdu1 = fits.CompImageHDU(data=a.copy())
+        hdu2 = fits.CompImageHDU(data=a.copy())
+        # Previously the following line would throw a TypeError,
+        # now it should be identical to the integer bzero case
+        hdu1.scale('int16', bzero=99.0)
+        hdu2.scale('int16', bzero=99)
+        assert np.allclose(hdu1.data, hdu2.data)
+
+        def test_scale_back_compressed_uint_assignment(self):
```
Comment (astrofrog) [inline astropy/io/fits/tests/test_image.py]: The indent for this test seems wrong

Comment (mwcraig) [inline astropy/io/fits/tests/test_image.py]: Good catch, fixed.

**Comment by astrofrog** [comment]
This is straightforward, so I'll go ahead and merge - thanks @bamford and @mwcraig!

**Comment by astrofrog** [comment]
Changelog fixed in https://github.com/astropy/astropy/commit/1c77c1ff69eaa019fd366d16cb09197e4b45256b

**Pull request merged**